dotnet/src/SemanticKernel/AI/OpenAI/Clients/RetryHandler.cs

```diff
@@ -0,0 +1,156 @@
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.SemanticKernel.AI.OpenAI.Clients;
+
+internal class RetryHandler: DelegatingHandler
+{
+    private const string Retry_After = "Retry-After";
+    private const int MaxRetries = 10;
+
+    protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+    {
+        int retryCount = 0;
+
+        while (true)
+        {
+            // Throw an exception if we've requested to cancel the operation
+            cancellationToken.ThrowIfCancellationRequested();
+
+            // Issue the request
+            HttpResponseMessage response = await base.SendAsync(request, cancellationToken);
+
+            // If the request does not require a retry then we're done
+            if (!ShouldRetry(response.StatusCode))
+            {
+                return response;
+            }
+
+            // Drain response content to free connections. Need to perform this
+            // before retry attempt and before the TooManyRetries ServiceException.
+            if (response.Content != null)
+            {                
+#if NET5_0_OR_GREATER
+                await response.Content.ReadAsByteArrayAsync(cancellationToken).ConfigureAwait(false);
+#else
+                await response.Content.ReadAsByteArrayAsync().ConfigureAwait(false);
+#endif
+            }
+
+            // Safety measure to not keep retrying forever
+            if (retryCount >= MaxRetries)
+            {
+                throw new AIException(AIException.ErrorCodes.UnknownError,
+                    $"Request reached it's max retry count of {retryCount}");
+            }
+
+            // Prepare Delay task configured with the delay time from response's Retry-After header
+            var waitTime = CalculateWaitTime(response);
+            Task delay = Task.Delay(waitTime, cancellationToken);
+
+            // Clone request with CloneAsync before retrying
+            // Do not dispose this request as that breaks the request cloning
+#pragma warning disable CA2000
+            request = await CloneAsync(request);
+#pragma warning restore CA2000
+
+            // Increase retryCount
+            retryCount++;
+
+            // Delay time based upon Retry-After header
+            await delay;
+        }
+    }
+
+    private static TimeSpan CalculateWaitTime(HttpResponseMessage response)
+    {
+        // Default delay, in case the retry-after header data is corrupt
+        double delayInSeconds = 10;
+
+        if (response != null && response.Headers.TryGetValues(Retry_After, out IEnumerable<string> values))
+        {
+            // Can we use the provided retry-after header?
+            string retryAfter = values.First();
+            if (int.TryParse(retryAfter, out int delaySeconds))
```

Comment:
The Retry-After header may have two values: 
- Date - the date to retry after
- Delta - time to wait for before next retry
Both should be respected because the retry handler is generic and can used by any potential service client.
![image](https://user-images.githubusercontent.com/68852919/222592056-787d4e94-9209-45da-b7bd-6e32ac0451ce.png)